src/core/utils/components/VirtualScroll.tsx

```diff
@@ -0,0 +1,232 @@
+  const onScrollRef = React.useRef<(e: Event) => void>();
+
+  // Find scrollable parent
+  // Needed only on init
+  React.useLayoutEffect(() => {
+    const scrollableParent = getScrollableParent(childrenParentRef.current);
+    scrollContainer.current = scrollableParent;
+  }, []);
+
+  const visibleChildren = React.useMemo(() => {
+    const arr = [];
+    const endIndex = Math.min(
+      childrenLength,
+      startNode + visibleNodeCount + renderAdditional,
+    );
+    for (let i = startNode; i < endIndex; i++) {
+      arr.push(itemRenderer(i));
+    }
+    return arr;
+  }, [
+    childrenLength,
+    itemRenderer,
+    renderAdditional,
+    startNode,
+    visibleNodeCount,
+  ]);
+
+  // Get child height when children available
+  React.useLayoutEffect(() => {
+    if (!childrenParentRef.current || !visibleChildren.length) {
+      return;
+    }
+    let heightSum = 0;
+    for (let i = startNode; i < visibleChildren.length + startNode; i++) {
+      const htmlElement = childrenParentRef.current.children.item(
+        Math.max(0, i - startNode),
+      ) as HTMLElement;
+      const elementHeight = getElementHeight(htmlElement);
+      heightSum += Number(elementHeight.toFixed(2));
+    }
+    const elementHeight = heightSum / visibleChildren.length;
+    childHeight.current = Math.ceil(elementHeight);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [visibleChildren.length]);
+
+  const onScroll = React.useCallback(
+    (e: Event) => {
+      const target =
+        scrollContainer.current !== document.body
+          ? (e.target as HTMLElement)
+          : (document.scrollingElement as HTMLElement);
+      if (!target) {
+        return;
+      }
+      const start = getCountOfNodesInHeight(
+        childHeight.current,
+        target.scrollTop,
+      );
+      setStartNode(start);
+      setVisibleNodeCount(
+        getVisibleNodeCount(childHeight.current, start, childrenLength, target),
+      );
+      setTranslateY(getTranslateValue(childHeight.current, start));
+    },
+    [childrenLength],
+  );
+
+  const removeScrollListener = React.useCallback(() => {
+    if (!onScrollRef.current) {
+      return;
+    }
+    !scrollContainer.current || scrollContainer.current === document.body
+      ? document.removeEventListener('scroll', onScrollRef.current)
+      : scrollContainer.current.removeEventListener(
+          'scroll',
+          onScrollRef.current,
+        );
+  }, []);
+
+  // Add event listener to the scrollable container.
+  React.useLayoutEffect(() => {
+    removeScrollListener();
+    onScrollRef.current = onScroll;
+    if (!scrollContainer.current || scrollContainer.current === document.body) {
+      document.addEventListener('scroll', onScroll);
+    } else {
+      scrollContainer.current.addEventListener('scroll', onScroll);
+    }
+    return removeScrollListener;
+  }, [onScroll, removeScrollListener]);
+
+  React.useLayoutEffect(() => {
```

Comment:
What it does? Is it for lazy-loading in chunks?

Reply:
It recalculates everything whenever children changes

Reply:
Isn't it doing the same as onScroll?

Reply:
well, it recalculates same things, yes

Reply:
So maybe don't duplicate code?

Reply:
stopped duplication